compiler/src/dotty/tools/dotc/transform/CheckShadowing.scala

```diff
@@ -0,0 +1,314 @@
+    val fresh = ctx.fresh.setProperty(_key, data)
+    shadowingDataApply(sd => sd.registerRootImports())(using fresh)
+
+  // Reporting on traversal's end
+  override def transformUnit(tree: tpd.Tree)(using Context): tpd.Tree =
+    shadowingDataApply(sd =>
+      reportShadowing(sd.getShadowingResult)
+    )
+    tree
+
+  // MiniPhase traversal :
+
+  override def prepareForPackageDef(tree: tpd.PackageDef)(using Context): Context =
+    shadowingDataApply(sd => sd.inNewScope())
+    ctx
+
+  override def prepareForTemplate(tree: tpd.Template)(using Context): Context =
+    shadowingDataApply(sd => sd.inNewScope())
+    ctx
+
+  override def prepareForBlock(tree: tpd.Block)(using Context): Context =
+    shadowingDataApply(sd => sd.inNewScope())
+    ctx
+
+  override def prepareForOther(tree: tpd.Tree)(using Context): Context =
+    importTraverser.traverse(tree)
+    ctx
+
+  override def prepareForValDef(tree: tpd.ValDef)(using Context): Context =
+    shadowingDataApply(sd =>
+      sd.registerPrivateShadows(tree)
+    )
+
+  override def prepareForTypeDef(tree: tpd.TypeDef)(using Context): Context =
+    if tree.symbol.isAliasType then // if alias, the parent is the current symbol
+      nestedTypeTraverser(tree.symbol).traverse(tree.rhs)
+    if tree.symbol.is(Param) then // if param, the parent is up
+      val owner = tree.symbol.owner
+      val parent = if (owner.isConstructor) then owner.owner else owner
+      nestedTypeTraverser(parent).traverse(tree.rhs)(using ctx.outer)
+      shadowingDataApply(sd => sd.registerCandidate(parent, tree))
+    else
+      ctx
+
+
+  override def transformPackageDef(tree: tpd.PackageDef)(using Context): tpd.Tree =
+    shadowingDataApply(sd => sd.outOfScope())
+    tree
+
+  override def transformBlock(tree: tpd.Block)(using Context): tpd.Tree =
+    shadowingDataApply(sd => sd.outOfScope())
+    tree
+
+  override def transformTemplate(tree: tpd.Template)(using Context): tpd.Tree =
+    shadowingDataApply(sd => sd.outOfScope())
+    tree
+
+  override def transformTypeDef(tree: tpd.TypeDef)(using Context): tpd.Tree =
+    if tree.symbol.is(Param) &&  isValidTypeParamOwner(tree.symbol.owner) then // Do not register for constructors the work is done for the Class owned equivalent TypeDef
+      shadowingDataApply(sd => sd.computeTypeParamShadowsFor(tree.symbol.owner)(using ctx.outer))
+    if tree.symbol.isAliasType then // No need to start outer here, because the TypeDef reached here it's already the parent
+      shadowingDataApply(sd => sd.computeTypeParamShadowsFor(tree.symbol)(using ctx))
+    tree
+
+  // Helpers :
+  private def isValidTypeParamOwner(owner: Symbol)(using Context): Boolean =
+    !owner.isConstructor && !owner.is(Synthetic) && !owner.is(Exported)
+
+  private def reportShadowing(res: ShadowingData.ShadowResult)(using Context): Unit =
+    res.warnings.sortBy(w => (w.pos.line, w.pos.startPos.column))(using Ordering[(Int, Int)]).foreach { s =>
+      s match
```

Comment:
When you pass lambda with pattern matching you can just skip the `s => s match` part